src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -155,7 +155,17 @@ public int getTotalUserCount(){
     for (Entity message : results.asList(FetchOptions.Builder.withLimit(1000))){
       users.add(message.getProperty("user"));
     }
-    System.out.println(users);
     return users.size();
   }
+
+  /** Returns the average message length. */
+  public double getAverageMessageLength(){
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+    Integer sum = 0;
+    for (Entity message : results.asList(FetchOptions.Builder.withLimit(1000))){
```

Comment:
take a look at the summary for the `PreparedQuery` javadoc, the two calls here may not be as efficient.

Reply:
Not super sure what you mean here, I changed line 169 to use `.countEntities()` instead of converting to a list, is that what you meant?

Reply:
The relevant part is "A PreparedQuery does not cache results. Each use of PreparedQuery results in a new trip to the datastore." 

Therefore, countEntities and asList will each go to the database which is slow. We should either assign `results.asList(FetchOptions.Builder.withDefaults()` to a variable and re-use that, or just keep counting in the for loop.

Reply:
Gotcha, thanks for the super quick feedback!